cmd/promoted-image-governor/main.go

```diff
@@ -0,0 +1,255 @@
+	fs.StringVar(&opts.openshiftMappingDir, "openshift-mapping-dir", "", "Path to the openshift mapping directory")
+	fs.StringVar(&opts.openshiftMappingConfigPath, "openshift-mapping-config", "", "Path to the openshift mapping config file")
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatal("could not parse input")
+	}
+	opts.ignoredImageStreamTags = opts.ignoredImageStreamTagsRaw.StringSet()
+	if opts.openshiftMappingConfigPath != "" {
+		c, err := loadOpenshiftMappingConfig(opts.openshiftMappingConfigPath)
+		if err != nil {
+			logrus.WithError(err).Fatal("could not load openshift mapping config")
+		}
+		opts.openshiftMappingConfig = c
+	}
+	return opts
+}
+
+func (o *options) validate() error {
+	if o.ciOperatorconfigPath == "" {
+		return fmt.Errorf("--ci-operator-config-path must be set")
+	}
+	if (o.openshiftMappingDir == "") != (o.openshiftMappingConfigPath == "") {
+		return fmt.Errorf("--openshift-mapping-dir and --openshift-mapping-config must be set together")
+	}
+	return nil
+}
+
+func addSchemes() error {
+	if err := imagev1.AddToScheme(scheme.Scheme); err != nil {
+		return fmt.Errorf("failed to add imagev1 to scheme: %w", err)
+	}
+	return nil
+}
+
+func validatePromotedTags(tags []api.ImageStreamTagReference) error {
+	tagCount := map[api.ImageStreamTagReference]int{}
+	for _, tag := range tags {
+		if _, ok := tagCount[tag]; ok {
+			tagCount[tag] = tagCount[tag] + 1
+		} else {
+			tagCount[tag] = 1
+		}
+	}
+	dups := sets.NewString()
+	for tag, c := range tagCount {
+		if c > 1 {
+			dups.Insert(fmt.Sprintf("%s/%s:%s", tag.Namespace, tag.Name, tag.Tag))
+		}
+	}
+	if dups.Len() > 0 {
+		return fmt.Errorf("found tags promoted more than once: %s", dups.List())
+	}
+	return nil
+}
+
+func tagsToDelete(ctx context.Context, client ctrlruntimeclient.Client, promotedTags []api.ImageStreamTagReference, toIgnore sets.String) (sets.String, error) {
+	tagsToCheck := sets.NewString()
+	var errs []error
+	imageStreamsWithPromotedTags := sets.NewString()
+
+	for _, promotedTag := range promotedTags {
+		namespacedImageStream := fmt.Sprintf("%s/%s", promotedTag.Namespace, promotedTag.Name)
+		if imageStreamsWithPromotedTags.Has(namespacedImageStream) {
+			continue
+		}
+		imageStreamsWithPromotedTags.Insert(namespacedImageStream)
+		imagestream := &imagev1.ImageStream{}
+		if err := client.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: promotedTag.Namespace, Name: promotedTag.Name}, imagestream); err != nil {
+			if !kerrors.IsNotFound(err) {
+				errs = append(errs, fmt.Errorf("could not get image steam %s in namespace %s", promotedTag.Name, promotedTag.Namespace))
+			}
+			continue
+		}
+		for _, tag := range imagestream.Status.Tags {
+			tagsToCheck.Insert(fmt.Sprintf("%s/%s:%s", promotedTag.Namespace, promotedTag.Name, tag.Tag))
```

Comment:
for sanity, use the imagestream's name and namespace for populating this

Reply:
Done.